lib/block_view.go

```diff
@@ -3593,6 +3648,139 @@ func (bav *UtxoView) _connectTransaction(txn *MsgDeSoTxn, txHash *BlockHash,
 	return utxoOpsForTxn, totalInput, totalOutput, fees, nil
 }
 
+func (bav *UtxoView) ValidateTransactionNonce(txn *MsgDeSoTxn, blockHeight uint64) error {
+	if txn == nil || txn.TxnNonce == nil {
+		return fmt.Errorf("ValidateTransactionNonce: Nonce or txn is nil for public key %v",
+			PkToStringBoth(txn.PublicKey))
+	}
+
+	if blockHeight > txn.TxnNonce.ExpirationBlockHeight {
+		return errors.Wrapf(RuleErrorNonceExpired,
+			"ValidateTransactionNonce: Nonce %s has expired for public key %v",
+			txn.TxnNonce.String(), PkToStringBoth(txn.PublicKey))
+	}
+	pkidEntry := bav.GetPKIDForPublicKey(txn.PublicKey)
+	if pkidEntry == nil || pkidEntry.isDeleted {
+		return fmt.Errorf(
+			"ValidateTransactionNonce: PKID for public key %s does not exist",
+			PkToString(txn.PublicKey, bav.Params))
+	}
+
+	nonce, err := bav.GetTransactorNonceEntry(txn.TxnNonce, pkidEntry.PKID)
+	if err != nil {
+		return errors.Wrapf(err,
+			"ValidateTransactionNonce: Problem getting transaction nonce entry for nonce %s and PKID %v",
+			txn.TxnNonce.String(), pkidEntry.PKID)
+	}
+	if nonce != nil && !nonce.isDeleted {
+		return errors.Wrapf(RuleErrorReusedNonce,
+			"ValidateTransactionNonce: Nonce %s has already been used for PKID %v",
+			txn.TxnNonce.String(), pkidEntry.PKID)
+	}
+	return nil
+}
+
+// _connectFailingTransaction is used to process the fee and burn associated with the user submitting a failing transaction.
+// A failing transaction passes formatting validation, yet fails connecting to the UtxoView. This can happen for a number
+// of reasons, such as insufficient DESO balance, wrong public key, etc. With Revolution's Fee-Time block ordering, these
+// failing transactions are included in the blocks and their fees are burned. Initially blocks will include full transactions,
+// and in the future iterations blocks will only include the failing transaction hashes. This is done to prevent spamming
+// attacks on the network with seemingly valid transactions that never commit and just end up occupying the block space.
+// In addition, a major part of the effective fees of this transaction is burned with BMF. This makes spam attacks economically
+// disadvantageous. Attacker's funds are burned, to the benefit of everyone else on the network. BMF algorithm also computes
+// a utility fee, which is distributed to the block producer.
```

Comment:
Don't put future speculation directly in comments. If you want, it can be added as a TODO underneath so that it's clear that we may not get to it.